File: py/src/imbrace/resources/boards.py
from typing import Any, Dict, Optional
from ..http import HttpTransport, AsyncHttpTransport


class BoardsResource:
    """Boards / Knowledge Hub / External Drive — Sync. Routes via gateway/data-board."""

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    # --- Boards ---
    def list(self, limit: int = 20, skip: int = 0) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards", params={"limit": limit, "skip": skip}).json()

    def get(self, board_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards/{board_id}").json()

    def get_by_contact(self, contact_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards/by-contact/{contact_id}").json()

    def create(self, name: str, description: Optional[str] = None) -> Dict[str, Any]:
        body: Dict[str, Any] = {"name": name}
        if description:
            body["description"] = description
        return self._http.request("POST", f"{self._base}/boards", json=body).json()

    def update(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/boards/{board_id}", json=body).json()

    def delete(self, board_id: str) -> None:
        self._http.request("DELETE", f"{self._base}/boards/{board_id}")

    def reorder(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/_order", json=body).json()

    def export_csv(self, board_id: str, params: Optional[Dict[str, str]] = None) -> str:
        return self._http.request("GET", f"{self._base}/boards/{board_id}/export_csv", params=params or {}).text

    def import_csv(self, board_id: str, files: Any) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/import_csv", files=files).json()

    def import_excel(self, board_id: str, files: Any) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/import_excel", files=files).json()

    def get_import_progress(self, board_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards/{board_id}/import_progress").json()

    def upload_board_file(self, files: Any) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/_fileupload", files=files).json()

    # --- Fields ---
    def create_field(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/fields", json=body).json()

    def update_field(self, board_id: str, field_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/boards/{board_id}/fields/{field_id}", json=body).json()

    def delete_field(self, board_id: str, field_id: str) -> None:
        self._http.request("DELETE", f"{self._base}/boards/{board_id}/fields/{field_id}")

    def reorder_fields(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/fields/reorder", json=body).json()

    def bulk_update_fields(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/boards/{board_id}/fields/bulk", json=body).json()

    # --- Items ---
    def list_items(self, board_id: str, limit: int = 20, skip: int = 0) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards/{board_id}/items",
                                  params={"limit": limit, "skip": skip}).json()

    def get_item(self, board_id: str, item_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards/{board_id}/items/{item_id}").json()

    def create_item(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/items", json=body).json()

    def update_item(self, board_id: str, item_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/boards/{board_id}/items/{item_id}", json=body).json()

    def delete_item(self, board_id: str, item_id: str) -> None:
        self._http.request("DELETE", f"{self._base}/boards/{board_id}/items/{item_id}")

    def bulk_delete_items(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("DELETE", f"{self._base}/boards/{board_id}/items/bulk-delete", json=body).json()

    def check_conflict(self, board_id: str, item_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/items/{item_id}/_is_conflicted", json=body).json()

    def get_related_items(self, board_id: str, item_id: str, related_board_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards/{board_id}/items/{item_id}/related/{related_board_id}").json()

    def link_items(self, board_id: str, item_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/items/{item_id}/related", json=body).json()

    def unlink_items(self, board_id: str, item_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("DELETE", f"{self._base}/boards/{board_id}/items/{item_id}/related", json=body).json()

    # --- Search ---
    def search(self, board_id: str, q: Optional[str] = None, limit: int = 100, offset: int = 0) -> Dict[str, Any]:
        body: Dict[str, Any] = {"limit": limit, "offset": offset}
        if q:
            body["q"] = q
        return self._http.request("POST", f"{self._base}/search/{board_id}", json=body).json()

    # --- Segmentation ---
    def list_segments(self, board_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/boards/{board_id}/segmentation").json()

    def create_segment(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/boards/{board_id}/segmentation", json=body).json()

    def update_segment(self, board_id: str, segment_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/boards/{board_id}/segmentation/{segment_id}", json=body).json()

    def delete_segment(self, board_id: str, segment_id: str) -> None:
        self._http.request("DELETE", f"{self._base}/boards/{board_id}/segmentation/{segment_id}")

    # --- Folders ---
    def search_folders(self, organization_id: str, q: Optional[str] = None) -> Dict[str, Any]:
        params: Dict[str, str] = {"organization_id": organization_id}
        if q:
            params["q"] = q
        return self._http.request("GET", f"{self._base}/folders/search", params=params).json()

    def get_folder(self, folder_id: str, recursive: Optional[bool] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if recursive is not None:
            params["recursive"] = str(recursive).lower()
        return self._http.request("GET", f"{self._base}/folders/{folder_id}", params=params).json()

    def create_folder(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/folders", json=body).json()

    def update_folder(self, folder_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/folders/{folder_id}", json=body).json()

    def delete_folders(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/folders/delete", json=body).json()

    def get_folder_contents(self, folder_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/folders/{folder_id}/contents").json()

    # --- Files ---
    def search_files(self, folder_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/files/search", params={"folder_id": folder_id}).json()

    def get_file(self, file_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/files/{file_id}").json()

    def create_file(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/files", json=body).json()

    def upload_file(self, files: Any) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/files/upload", files=files).json()

    def download_file(self, file_id: str) -> Any:
        return self._http.request("GET", f"{self._base}/files/{file_id}/download")

    def update_file(self, file_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/files/{file_id}", json=body).json()

    def delete_files(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/files/delete", json=body).json()

    def generate_ai_tags(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/ai/tag-generation", json=body).json()

    def get_link_preview(self, url: str) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._base}/link_preview/getWebsiteInfo", json={"url": url}).json()

    # --- External Drive ---
    def initiate_drive_auth(self, drive_type: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/auth/{drive_type}/initiate").json()

    def list_drive_folders(self, drive_type: str, params: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/{drive_type}/folders", params=params or {}).json()

    def list_drive_files(self, drive_type: str, params: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/{drive_type}/files", params=params or {}).json()

    def download_drive_file(self, drive_type: str, params: Optional[Dict[str, str]] = None) -> Any:
        return self._http.request("GET", f"{self._base}/{drive_type}/files/download", params=params or {})

    def get_onedrive_session_status(self) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/auth/onedrive/files/session/status").json()


class AsyncBoardsResource:
    """Boards / Knowledge Hub / External Drive — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    async def list(self, limit: int = 20, skip: int = 0) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._base}/boards", params={"limit": limit, "skip": skip})
        return res.json()

    async def get(self, board_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._base}/boards/{board_id}")
        return res.json()

    async def create(self, name: str, description: Optional[str] = None) -> Dict[str, Any]:
        body: Dict[str, Any] = {"name": name}
        if description:
            body["description"] = description
        res = await self._http.request("POST", f"{self._base}/boards", json=body)
        return res.json()

    async def update(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PUT", f"{self._base}/boards/{board_id}", json=body)
        return res.json()

    async def delete(self, board_id: str) -> None:
        await self._http.request("DELETE", f"{self._base}/boards/{board_id}")

    async def list_items(self, board_id: str, limit: int = 20, skip: int = 0) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._base}/boards/{board_id}/items",
                                       params={"limit": limit, "skip": skip})
        return res.json()

    async def get_item(self, board_id: str, item_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._base}/boards/{board_id}/items/{item_id}")
        return res.json()

    async def create_item(self, board_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._base}/boards/{board_id}/items", json=body)
        return res.json()

    async def update_item(self, board_id: str, item_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PUT", f"{self._base}/boards/{board_id}/items/{item_id}", json=body)
        return res.json()

    async def delete_item(self, board_id: str, item_id: str) -> None:
        await self._http.request("DELETE", f"{self._base}/boards/{board_id}/items/{item_id}")

    async def search(self, board_id: str, q: Optional[str] = None, limit: int = 100, offset: int = 0) -> Dict[str, Any]:
        body: Dict[str, Any] = {"limit": limit, "offset": offset}
        if q:
            body["q"] = q
        res = await self._http.request("POST", f"{self._base}/search/{board_id}", json=body)
        return res.json()

    async def export_csv(self, board_id: str) -> str:
        res = await self._http.request("GET", f"{self._base}/boards/{board_id}/export_csv")
        return res.text

    async def upload_file(self, files: Any) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._base}/files/upload", files=files)
        return res.json()

    async def get_folder(self, folder_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._base}/folders/{folder_id}")
        return res.json()

    async def search_folders(self, organization_id: str, q: Optional[str] = None) -> Dict[str, Any]:
        params: Dict[str, str] = {"organization_id": organization_id}
        if q:
            params["q"] = q
        res = await self._http.request("GET", f"{self._base}/folders/search", params=params)
        return res.json()

File: ts/src/resources/boards.ts
import { HttpTransport } from "../http.js"

export class BoardsResource {
  /**
   * @param base - data-board base URL (gateway/data-board)
   *   Không có version prefix — path trực tiếp.
   */
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  // ─── Boards ──────────────────────────────────────────────────────────────────

  async list(params?: { limit?: number; skip?: number }) {
    const url = new URL(`${this.base}/boards`)
    if (params?.limit)  url.searchParams.set("limit", String(params.limit))
    if (params?.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async get(boardId: string) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}`, { method: "GET" }).then(r => r.json())
  }

  async getByContact(contactId: string) {
    return this.http.getFetch()(`${this.base}/boards/by-contact/${contactId}`, { method: "GET" }).then(r => r.json())
  }

  async create(body: { name: string; description?: string }) {
    return this.http.getFetch()(`${this.base}/boards`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async update(boardId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async delete(boardId: string): Promise<void> {
    await this.http.getFetch()(`${this.base}/boards/${boardId}`, { method: "DELETE" })
  }

  async reorder(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/_order`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async exportCsv(boardId: string, params?: Record<string, string>) {
    const url = new URL(`${this.base}/boards/${boardId}/export_csv`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.text())
  }

  async importCsv(boardId: string, body: FormData) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/import_csv`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  async importExcel(boardId: string, body: FormData) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/import_excel`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  async getImportProgress(boardId: string) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/import_progress`, { method: "GET" }).then(r => r.json())
  }

  // ─── Fields ──────────────────────────────────────────────────────────────────

  async createField(boardId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/fields`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateField(boardId: string, fieldId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/fields/${fieldId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteField(boardId: string, fieldId: string): Promise<void> {
    await this.http.getFetch()(`${this.base}/boards/${boardId}/fields/${fieldId}`, { method: "DELETE" })
  }

  async reorderFields(boardId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/fields/reorder`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async bulkUpdateFields(boardId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/fields/bulk`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ─── Items (records) ─────────────────────────────────────────────────────────

  async listItems(boardId: string, params?: { limit?: number; skip?: number }) {
    const url = new URL(`${this.base}/boards/${boardId}/items`)
    if (params?.limit)  url.searchParams.set("limit", String(params.limit))
    if (params?.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getItem(boardId: string, itemId: string) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items/${itemId}`, { method: "GET" }).then(r => r.json())
  }

  async createItem(boardId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateItem(boardId: string, itemId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items/${itemId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteItem(boardId: string, itemId: string): Promise<void> {
    await this.http.getFetch()(`${this.base}/boards/${boardId}/items/${itemId}`, { method: "DELETE" })
  }

  async bulkDeleteItems(boardId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items/bulk-delete`, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async checkConflict(boardId: string, itemId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items/${itemId}/_is_conflicted`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async getRelatedItems(boardId: string, itemId: string, relatedBoardId: string) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items/${itemId}/related/${relatedBoardId}`, { method: "GET" }).then(r => r.json())
  }

  async linkItems(boardId: string, itemId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items/${itemId}/related`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async unlinkItems(boardId: string, itemId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/items/${itemId}/related`, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ─── Search ──────────────────────────────────────────────────────────────────

  async search(boardId: string, body: { q?: string; limit?: number; offset?: number }) {
    return this.http.getFetch()(`${this.base}/search/${boardId}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ─── Segmentation ────────────────────────────────────────────────────────────

  async listSegments(boardId: string) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/segmentation`, { method: "GET" }).then(r => r.json())
  }

  async createSegment(boardId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/segmentation`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateSegment(boardId: string, segmentId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/boards/${boardId}/segmentation/${segmentId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteSegment(boardId: string, segmentId: string): Promise<void> {
    await this.http.getFetch()(`${this.base}/boards/${boardId}/segmentation/${segmentId}`, { method: "DELETE" })
  }

  // ─── Knowledge Hub (folders/files via data-board) ────────────────────────────

  async searchFolders(params: { organizationId: string; q?: string }) {
    const url = new URL(`${this.base}/folders/search`)
    url.searchParams.set("organization_id", params.organizationId)
    if (params.q) url.searchParams.set("q", params.q)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getFolder(folderId: string, params?: { recursive?: boolean }) {
    const url = new URL(`${this.base}/folders/${folderId}`)
    if (params?.recursive !== undefined) url.searchParams.set("recursive", String(params.recursive))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async createFolder(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/folders`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateFolder(folderId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/folders/${folderId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteFolders(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/folders/delete`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async searchFiles(params: { folderId: string }) {
    const url = new URL(`${this.base}/files/search`)
    url.searchParams.set("folder_id", params.folderId)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getFile(fileId: string) {
    return this.http.getFetch()(`${this.base}/files/${fileId}`, { method: "GET" }).then(r => r.json())
  }

  async createFile(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/files`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async uploadFile(body: FormData) {
    return this.http.getFetch()(`${this.base}/files/upload`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  async downloadFile(fileId: string) {
    return this.http.getFetch()(`${this.base}/files/${fileId}/download`, { method: "GET" })
  }

  async deleteFiles(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/files/delete`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async generateAiTags(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/ai/tag-generation`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async getLinkPreview(url: string) {
    return this.http.getFetch()(`${this.base}/link_preview/getWebsiteInfo`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ url }),
    }).then(r => r.json())
  }

  // ─── Board file upload ───────────────────────────────────────────────────────

  async uploadBoardFile(body: FormData) {
    return this.http.getFetch()(`${this.base}/boards/_fileupload`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  // ─── Folder contents ─────────────────────────────────────────────────────────

  async getFolderContents(folderId: string) {
    return this.http.getFetch()(`${this.base}/folders/${folderId}/contents`, { method: "GET" }).then(r => r.json())
  }

  // ─── File update ─────────────────────────────────────────────────────────────

  async updateFile(fileId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.base}/files/${fileId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ─── External Drive integration ──────────────────────────────────────────────

  async initiateDriveAuth(type: string) {
    return this.http.getFetch()(`${this.base}/auth/${type}/initiate`, { method: "GET" }).then(r => r.json())
  }

  async listDriveFolders(type: string, params?: Record<string, string>) {
    const url = new URL(`${this.base}/${type}/folders`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async listDriveFiles(type: string, params?: Record<string, string>) {
    const url = new URL(`${this.base}/${type}/files`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async downloadDriveFile(type: string, params?: Record<string, string>) {
    const url = new URL(`${this.base}/${type}/files/download`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" })
  }

  async getOneDriveSessionStatus() {
    return this.http.getFetch()(`${this.base}/auth/onedrive/files/session/status`, { method: "GET" }).then(r => r.json())
  }
}
